crates/bevy_reflect/src/serde/de.rs

```diff
@@ -12,375 +16,476 @@ pub trait DeserializeValue {
-                        registry: self.registry,
-                    })?;
-                    return Ok(Box::new(tuple));
-                }
-                type_fields::LIST => {
-                    let _type_name = type_name
-                        .take()
-                        .ok_or_else(|| de::Error::missing_field(type_fields::TYPE))?;
-                    let list = map.next_value_seed(ListDeserializer {
-                        registry: self.registry,
-                    })?;
-                    return Ok(Box::new(list));
-                }
-                type_fields::ARRAY => {
-                    let _type_name = type_name
-                        .take()
-                        .ok_or_else(|| de::Error::missing_field(type_fields::TYPE))?;
-                    let array = map.next_value_seed(ArrayDeserializer {
-                        registry: self.registry,
-                    })?;
-                    return Ok(Box::new(array));
-                }
-                type_fields::ENUM => {
-                    let type_name = type_name
-                        .take()
-                        .ok_or_else(|| de::Error::missing_field(type_fields::TYPE))?;
-                    let mut dynamic_enum = map.next_value_seed(EnumDeserializer {
-                        registry: self.registry,
-                    })?;
-                    dynamic_enum.set_name(type_name);
-                    return Ok(Box::new(dynamic_enum));
-                }
-                type_fields::VALUE => {
-                    let type_name = type_name
-                        .take()
-                        .ok_or_else(|| de::Error::missing_field(type_fields::TYPE))?;
-                    let registration =
-                        self.registry.get_with_name(&type_name).ok_or_else(|| {
-                            de::Error::custom(format_args!(
-                                "No registration found for {}",
-                                type_name
-                            ))
-                        })?;
-                    let deserialize_reflect =
-                        registration.data::<ReflectDeserialize>().ok_or_else(|| {
-                            de::Error::custom(format_args!(
-                                "The TypeRegistration for {} doesn't have DeserializeReflect",
-                                type_name
-                            ))
-                        })?;
-                    let value = map.next_value_seed(DeserializeReflectDeserializer {
-                        reflect_deserialize: deserialize_reflect,
-                    })?;
-                    return Ok(value);
-                }
-                _ => return Err(de::Error::unknown_field(key.as_str(), &[])),
-            }
-        }
-
-        Err(de::Error::custom("Maps in this location must have the \'type\' field and one of the following fields: \'map\', \'seq\', \'value\'"))
-    }
+        let type_name = map
+            .next_key::<&str>()?
```

Comment:
This change proves problematic, as `&str` will panic when deserializing `String`.

Suggest to deserialize `String` directly instead as this will make the deserializer more robust, and is what the previous deserializer used.

This is particularly painful when deserializing from foreign `Deserializer` objects.